zookeeper-server/src/main/java/org/apache/zookeeper/server/admin/Commands.java

```diff
@@ -392,8 +392,12 @@ public CommandResponse run(ZooKeeperServer zkServer, Map<String, String> kwargs)
                 response.put("observer_master_id", ((ObserverZooKeeperServer)zkServer).getObserver().getLearnerMasterId());
             }
 
-            response.putAll(ServerMetrics.getAllValues());
-
+            ServerMetrics.getMetrics()
+                    .getMetricsProvider()
+                    .dump(
+                    (metric, value) -> {
```

Comment:
Is this ordered in a deterministic fashion or could values jump around depending on the time?

Reply:
@enixon

Currently the API does not specify any ordering requirement.
Do you think it is needed?

If it is needed I wonder if we have to declare a specific ordering (by key?) or should we impose that the relative order of keys should not change over time (I image that new keys can appear and some keys may disappear) ?

Personally I wouldn't set any constraint and leave the result up to the provider.

Please note that we are also not requiring that the same key can appear only once with this kind of API

This new implementation (DefaultMetricsProvider)  does not track any ordering.

Let me also add that maybe new Providers may not expose data to this 'commands' and they will have an empty 'dump' method (Prometheus will expose metrics using its own endpoint, some provider will use JMX...)

Reply:
For any automated system, it probably doesn't matter. I'm just thinking that a human operator would appreciate consistency between different iterations of the command - particularly if the whole list gets long. :)

Not a blocker in any sense, just a convenience factor thought.

Reply:
btw As metrics are created at boot and they don't change I expect the order not to change if you don't restart the server. At least for the DefaultMetricsProvider.


@enixon